<a href="https://colab.research.google.com/github/rifyalt/basic_python/blob/main/8_Data_Grouping.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Grouping

Dalam kesempatan sekarang kita akan membahas teknik-teknik grouping pada data.

Grouping adalah suatu teknik memisahkan data berdasarkan kriteria tertentu. untuk melakukan grouping di pandas kita menggunakan suatu method yang bernama groupby(). definisi abstraknya dari grouping adalah kita melakukan mapping suatu data terhadap suatu group.

Ketika melakukan grouping ada beberapa proses yang terjadi secara berurutan, yaitu :

*   Splitting : Memisahkan data kedalam suatu group berdasarkan kriteria tertentu.
*   Applying : Melakukan suatu operasi terhadap sekumpulan data di group-group tersebut.
*   Combining : Menggabungkan data menjadi suatu struktur baru

Langsung saja kita akan melihat bagaimana process split-apply-combine.Kita akan menggunakan data sederhana untuk tujuan pembelajaran.

In [ ]:
import numpy as np
import pandas as pd
df = pd.DataFrame([('bird', 'Falconiformes', 389.0),
                           ('bird', 'Psittaciformes', 24.0),
                           ('mammal', 'Carnivora', 80.2),
                           ('mammal', 'Primates', np.nan),
                           ('mammal', 'Carnivora', 58)],
                          index=['falcon', 'parrot', 'lion', 'monkey', 'leopard'],
                             columns=('class', 'order', 'max_speed'))

df

class           order  max_speed
falcon     bird   Falconiformes      389.0
parrot     bird  Psittaciformes       24.0
lion     mammal       Carnivora       80.2
monkey   mammal        Primates        NaN
leopard  mammal       Carnivora       58.0

In [ ]:
# Mari kita praktikan method groupby terhadap data tersebut. misalkan kita group data tersebut berdasarkan class.

df.groupby('class')

Coba kalian perhatikan, pada tahap ini kita baru sampai pada process splitting berdasarkan kategori ‘class’. hasil dari groupby ini hanya sebuah objek GroupBy. objek groupby ini adalah representasi dari suatu hasil penggabungan data berdasarkan group class, namun tidak menghasilkan nilai apapun sebelum kita memberikan suatu operasi terhadap hasil dari setiap grouping tersebut.

Prosess melakukan operasi terhadap sekumpulan data yang telah di groupkan adalah process applying. kita bisa menggunakan berbagai jenis operasi aggregasi terhadap objek groupby tersebut. ada operasi yang telah di sediakan sebagai operasi bawaan dari pandas seperti mean(), min(), max(), dan lain sebagainya. dapat pula menggunakan custom operasi tergantung apa yang kita inginkan. mari kita coba praktekan salah satu jenis operasi kepada objek groupby tersebut.

In [ ]:
df.groupby('class').mean()

max_speed
class            
bird        206.5
mammal       69.1

Coba kalian perhatikan. kita telah berhasil menggunakan groupby dan memberikan hasil. jenis operasi yang di apply terhadap setiap group adalah rata-rata. dan hasilnya di combine menjadi struktur data baru. dengan begini kita menhasilkan suatu hasil analisis sederhana yaitu bahwa binatang dengan class bird memiliki kecepatan maksimal rata-rata sebesar 206.5 dan binatang class mamalia sebesar 69.1.

Kita juga dapat mengambil lebih banyak data dengan cara:

In [ ]:
df.groupby(['class', 'order'])[['max_speed']].agg(['mean', 'sum'])

max_speed       
                           mean    sum
class  order                          
bird   Falconiformes      389.0  389.0
       Psittaciformes      24.0   24.0
mammal Carnivora           69.1  138.2
       Primates             NaN    0.0

Beberapa Pengolahan Data yang Dapat Dilakukan, Silahkan kalian tonton video berikut:

*   [Pandas Group by](https://youtu.be/mSsLhrqv-aA)

